### PR TITLE
opt(gemm): reduce MXFP4 dense dispatch overhead

### PR DESCRIPTION
**Pull request by suachong**
## Summary

This PR carries the reviewed champion from the GPT-OSS-20B MXFP4 dense GEMM campaign on MI355X.

It:

- repairs the MXFP4 autotune cache and variant-race keys so they include the true padded-row contract and beta mode;
- forwards `beta_is_one` when compiling and racing TACCW/COOP variants, allowing beta=1 WGRAD shapes to select their real epilogue;
- memoizes pure shape/dtype/beta-derived launch metadata without using tensor identity or data pointers;
- avoids redundant device lookups, no-op contiguous conversions, and scale flattening in the Python launch path.

Only `primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py` changes. The final source also includes a Ruff-only line wrap relative to the measured champion.

## Campaign provenance

- MLPerf branch: `suachong/gpt-oss-mxfp4-dense-gemm-campaign`
- Workflow: https://github.com/AMD-MLPerf/mlperf-training/actions/runs/34728076127
- GPU: AMD Instinct MI355X / gfx950
- Image: `rocm/mlperf:small-llm-moe-pretraining-primus_training_6.0_2026-09-12-14-28-49_mi355x_dev`
- Image digest: `rocm/mlperf@sha256:19341a1fd236257de520bae7e6c854411f9c4f34fce114bb48d42ac56c41c2d0`
- Primus-Turbo baseline: `ad5431f077314ebc5e9a07fd08c69a94641c7da5`
- Campaign length: 20 rounds: 3 analyze, 13 optimize/replan, 2 review, plus orchestration
- Metric: arithmetic mean PFLOP/s across six production calls; higher is better
- Ruler: 20 warmups, 30 timed repetitions, three campaign repetitions
- Hard gates: SNR >=20 dB against FP32 matmul of dequantized operands and byte-identical output over two independent launches

## Production shapes

| Case | M | N | K | beta | Row stride |
|---|---:|---:|---:|---:|---:|
| qkv_fwd | 32768 | 5120 | 2880 | 0 | 1472 B |
| qkv_dgrad | 32768 | 2880 | 5120 | 0 | 2560 B |
| qkv_wgrad | 5120 | 2880 | 32768 | 1 | 16384 B |
| output_fwd | 32768 | 2880 | 4096 | 0 | 2048 B |
| output_dgrad | 32768 | 4096 | 2880 | 0 | 1472 B |
| output_wgrad | 2880 | 4096 | 32768 | 1 | 16384 B |

MXFP4 quantization and dequantization are outside the timed region. WGRAD includes the production beta=1 accumulation path. K=2880 retains the quantizer-produced 1472-byte row stride; no extra padding or caller-contract change was introduced.

## Scoring

| | mean PFLOP/s | Change |
|---|---:|---:|
| Baseline | 3.87790367 | -- |
| Autotune-key fix | 3.95231133 | +1.92% |
| Final champion | 3.98847883 | +2.85% |

The exact best recorded baseline and champion repetitions were:

| Case | Baseline PFLOP/s | Champion PFLOP/s | Champion ms | SNR dB |
|---|---:|---:|---:|---:|
| qkv_fwd | 3.812357 | 3.830498 | 0.25228 | 55.38 |
| qkv_dgrad | 4.009791 | 4.070596 | 0.23740 | 55.46 |
| qkv_wgrad | 4.151019 | 4.261961 | 0.22674 | 52.60 |
| output_fwd | 3.974736 | 4.093434 | 0.18886 | 55.44 |
| output_dgrad | 3.634293 | 3.889938 | 0.19874 | 55.39 |
| output_wgrad | 3.685226 | 3.784446 | 0.20428 | 52.59 |

Per-cell values above are individual best-score repetitions and should not be read as paired per-cell estimates on the shared runner. The more controlled round-4 experiment used 8 baseline and 11 candidate interleaved runs: median `mean_pf` rose 3.82865 -> 3.88944 (+1.59%), driven by qkv_wgrad +4.56% and output_wgrad +2.52%; the other four cases stayed within their measured noise bands. Kernel traces showed WGRAD scratch falling from 184/160 B to zero after the real beta=1 TACCW variants became selectable.

The host-dispatch cleanup was measured over 12 interleaved pairs: median 3.899981 -> 3.929990 (+0.769%), paired mean +0.837%, t=4.361, p approximately 0.0013, with 11/12 positive signs and no correctness failures.

## Correctness and transferability

- Every retained candidate run reported `ok=true` for all six cases.
- Determinism was 2/2 byte-identical launches for every case.
- SNR remained 52.59-55.46 dB, well above the 20 dB gate.
- The beta=1 reference includes the same nonzero accumulation buffer as the kernel.
- All six cases passed CUDA graph capture and five replays after the host-path cleanup.
- The new shape cache is keyed only by M/N/K, packed row extent, output dtype, and beta. It contains only scalar metadata and never keys on tensor identity.
- Final review found no benchmark, scoring, quantization, caller-contract, debug, or unrelated-file changes.

## Campaign findings

Kept:

1. The TACCW/COOP race was effectively dead because its lookup tuple did not match the 14-element key written by the sweep; the swallowed KeyError prevented the variants from being evaluated. The lookup now includes `k_real`, `row_bytes`, and the beta mode.
2. Beta=1 needs a distinct autotune/cache entry because its read-modify-write epilogue is a different kernel. Correctly racing it selected TACCW for both WGRAD shapes and COOP for output_wgrad.
3. Repeated launch-side shape calculations and no-op tensor transformations are measurable at this dispatch frequency. Shape-only memoization and avoiding redundant conversions produced a statistically supported additional gain.

Measured and reverted:

- Per-workgroup in-GEMM scale repacking was correct but did approximately 11x redundant work. It regressed qkv_dgrad by about 12% and output_fwd by about 8.6%. Reordering its loads with operand prefetch did not recover the loss.
- Python-level ROWSPLIT address hoists changed MLIR but generated byte-identical ISA because compiler CSE already removes them.
- Increasing A-refill depth measured -0.13%; widening the deep-pipeline autotune gate measured -0.75%. Both were reverted.
- The 8-wave aligned-path prototype failed register allocation on the production kernel even after isolating an unrelated COOP routing issue. The toy kernel remained useful for validating address formulas, but no production change was kept.
- Removing the second prologue wait/barrier caused nondeterminism on all adversarial trials for the two ROWSPLIT K=2880 cases. The full wait cannot simply be deleted; any follow-up must preserve LDS producer-consumer ordering.
- Split-K was not beneficial for output_wgrad in the measured regime: about 178 us versus 167 us for the settled path including reduction.
- Shared-runner/DVFS variation was material. Changes were accepted using repeated interleaved measurements and paired statistics rather than single-run deltas.

## Open follow-ups

The campaign left these measured leads for separate PRs:

1. Front-load the first phase-A G2S prefix for the four beta=0 cases, then place a precisely counted `s_waitcnt vmcnt(18) lgkmcnt(0)` plus barrier before the slot-1 reads. Validate first with the adversarial E8M0 toy/probe protocol.
2. Merge the nonredundant scale-preshuffle blocks and GEMM blocks into one launch using a block-uniform branch. Standalone FlyDSL probes confirmed that block-uniform `scf.if` with branch-local LDS and ping-pong barriers is byte-correct.
3. Redesign the ROWSPLIT ring rotation in raw ASM. The aligned-vs-ROWSPLIT ISA delta isolates 12 `v_add_u32`, 8 `v_mov_b32`, and 12 `s_mov_b32` instructions per steady-loop iteration.
4. Revisit cross-tile prime overlap after a dedicated toy-kernel proof; the fitted model attributes about 3.077 us per tile to fixed restart work and about 5.765 us per dispatch.

## Validation

- Full six-case campaign ruler, repeated and interleaved as described above
- Adversarial E8M0 exponent probes for synchronization-sensitive experiments
- CUDA graph capture/replay on all six production cases
- Final diff review: one file only, no debug/probe/dead variants
- `ruff check primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py`
- `ruff format --check primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py`
- Python AST parse
- `git diff --check`

### CODE REVIEW

File: primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py [hunk 1]
```diff
@@ -2822,7 +2888,7 @@ def _exec_split(ksplit, target=None, accum=False):
             ksplit=ksplit,
             out_fp16=out_fp16,
             n_tail=N % 256,
-            mn=_mxfp4_mn_specialise(M, N),
+            mn=_mn,
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py:3012]: Addressed in 686927ce. The split-K winner cache now includes device and beta_is_one, and both plain and split candidates are timed with accum=beta_is_one so the selected path includes the real beta=1 epilogue/add cost.

File: primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py [hunk 2]
```diff
@@ -2350,6 +2372,32 @@ def _mxfp4_mn_specialise(M, N):
...
+    collapses ~6 small per-call computations into one dict lookup."""
+    key = (M, N, K, Kb_a, out_dtype, beta)
+    rec = _MXFP4_SHAPE_CACHE.get(key)
+    if rec is None:
+        Kw = (K + 255) // 256 * 256  # loop + packed-scale extent
+        k_real = None if K == Kw else K  # None keeps the aligned shapes' launch key unchanged
+        row_bytes = None if Kb_a == K // 2 else Kb_a
+        out_fp16 = out_dtype == torch.float16
+        beta_is_one = beta == 1.0
+        mn = _mxfp4_mn_specialise(M, N)
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py]: Addressed in 686927ce. The shape record now keys on the operand device index, and the CU-count cache is a per-device map populated with get_device_properties(device_idx). The same explicit device is also used by the tile-specialisation and split-K candidate paths.

File: primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py [hunk 3]
```diff
@@ -2717,28 +2764,43 @@ def gemm_mxfp4_flydsl_kernel(
...
+    # only (never tensor identity), so a real shape change still misses and recomputes fresh.
+    Kw, _k_real, _row_b, out_fp16, beta_is_one, _mn = _mxfp4_shape_derived(M, N, K, Kb_a, out_dtype, beta)
+
+    device = a.device  # hoisted: reused below instead of 3 separate `a.device` property reads
+    # Pass the int index, not the device object: torch.cuda._utils._get_device_index() special-
+    # cases isinstance(dev, int) first and returns immediately, skipping the torch.device /
+    # jit.is_scripting dispatch chain that current_stream(device) or current_stream() still walk.
+    # Also correctness-neutral-or-better: pins the stream to the device the operands actually
+    # live on instead of whatever the ambient "current device" happens to be.
+    stream = torch.cuda.current_stream(device.index)
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py]: Addressed in 686927ce. Capture state is queried inside torch.cuda.device(device_idx), where device_idx comes from a.device. Autotune synchronization and event recording are also explicitly bound to that device/stream.

File: primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py [hunk 4]
```diff
@@ -2784,7 +2850,7 @@ def _exec_plain(target=None, accum=False):
             n_tail=N % 256,
             k_real=_k_real,
             row_bytes=_row_b,
-            mn=_mxfp4_mn_specialise(M, N),
+            mn=_mn,
         )
         # row_bytes must be in the artifact key too: one logical shape, two allocations, two kernels.
         at_key = (M, N, K, _row_b, gm, xcd, gn, _wlv, _elgk, _tw, _coop, out_fp16, accum)
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus_turbo/flydsl/gemm/gemm_mxfp4_kernel.py]: Addressed in 686927ce. All compiled-artifact reads and writes now use _mxfp4_artifact_key, including device, padded Kw, logical k_real, row_bytes, tuning parameters, ksplit, dtype, and beta. The final plain launch therefore reuses the exact entry compiled by autotune.

**Comment by suachong** [comment]
Review follow-up is in 686927ce. In addition to the four resolved inline threads, the suppressed `k_real` finding is addressed: `_MXFP4_CFG_CACHE` now keys on both padded `K` and logical `k_real`, plus the operand device. Local license, syntax, Ruff lint/format, and whitespace checks pass; the PR code-lint job is also green.

**Review by RuibinCheung — CHANGES_REQUESTED** [review]
I think this change doesn't make sense. In torch distributed training, 1 process owns 1 gpu. The process always map to one device.
